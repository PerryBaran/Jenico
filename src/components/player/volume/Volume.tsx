import React, { useState, useEffect, RefObject, ChangeEvent} from 'react';
import style from './volume.module.css';
import { mediumVolume, lowVolume, muteVolume } from '../../../media/icons/index';

function Volume(props: {audioRef: RefObject<HTMLAudioElement>}) {
    const { audioRef } = props
    const [volume, setVolume] = useState(0.5);
    const [icon, setIcon] = useState(mediumVolume);

    useEffect(() => {
        if (audioRef.current) {
            audioRef.current.volume = volume;
        }
    }, [volume, audioRef]);

    useEffect(() => {
        setIcon(()=> {
            if (volume === 0) {
                return muteVolume
            }
            if (volume < 0.5) {
                return lowVolume
            }
            return mediumVolume
        });
    }, [volume]);

    const changeVolume = (e: ChangeEvent<HTMLInputElement>) => {
        const currentVolume = Number(e.target.value)/100;
        setVolume(currentVolume);
    };

    return (
        <div className={style.container}>
                    <input
                        type='range'
                        name='volume'
                        min={0}
                        max={100}
                        defaultValue={volume * 100}
                        onChange={e => changeVolume(e)}/>
                    <button><img src={icon} alt='volume'/></button>
                </div>
    );
}

export default Volume;