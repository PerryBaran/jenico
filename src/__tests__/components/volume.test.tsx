import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Volume from "../../components/player/volume/Volume";

describe("Volume", () => {
  const validProps = {
    volume: 0.5,
    handleVolume: jest.fn(),
  };

  afterEach(cleanup);

  test("snapshot", () => {
    const { asFragment } = render(
      <Volume
        volume={validProps.volume}
        handleVolume={validProps.handleVolume}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("renders correctly, volume >= 0.5", () => {
    render(
      <Volume
        volume={validProps.volume}
        handleVolume={validProps.handleVolume}
      />
    );
    const input = screen.getByTestId("volume");

    expect(input).toHaveAttribute("min", "0");
    expect(input).toHaveAttribute("max", "100");
    expect(input).toHaveAttribute("value", `${validProps.volume * 100}`);
    expect(screen.getByAltText("volume")).toHaveAttribute(
      "src",
      "medium-volume.png"
    );
  });

  test("renders correctly, 0 < volume < 0.5", () => {
    const volume = 0.3;
    render(<Volume volume={volume} handleVolume={validProps.handleVolume} />);

    expect(screen.getByTestId("volume")).toHaveAttribute(
      "value",
      `${volume * 100}`
    );
    expect(screen.getByAltText("volume")).toHaveAttribute(
      "src",
      "low-volume.png"
    );
  });

  test("renders correctly, volume = 0", () => {
    const volume = 0;
    render(<Volume volume={volume} handleVolume={validProps.handleVolume} />);

    expect(screen.getByTestId("volume")).toHaveAttribute(
      "value",
      `${volume * 100}`
    );
    expect(screen.getByAltText("volume")).toHaveAttribute(
      "src",
      "volume-mute.png"
    );
  });

  test("input", () => {
    render(
      <Volume
        volume={validProps.volume}
        handleVolume={validProps.handleVolume}
      />
    );
    const newVolume = "0.7";

    expect(validProps.handleVolume).toBeCalledTimes(0);
    fireEvent.change(screen.getByTestId("volume"), {
      target: { value: newVolume },
    });
    expect(validProps.handleVolume).toBeCalledTimes(1);
    expect(validProps.handleVolume).toBeCalledWith(Number(newVolume) / 100);
  });
});
