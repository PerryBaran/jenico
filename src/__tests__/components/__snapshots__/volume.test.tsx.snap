// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Volume snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <input
      data-testid="volume"
      max="100"
      min="0"
      name="volume"
      type="range"
      value="50"
    />
    <img
      alt="volume"
      src="medium-volume.png"
    />
  </div>
</DocumentFragment>
`;
